controller/service.go

```diff
@@ -47,76 +52,82 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
-		lbIP = net.ParseIP(svc.Status.LoadBalancer.Ingress[0].IP)
+	for i := range svc.Status.LoadBalancer.Ingress {
+		lbIPs = append(lbIPs, net.ParseIP(svc.Status.LoadBalancer.Ingress[i].IP))
 	}
-	if lbIP == nil {
+
+	if len(lbIPs) == 0 {
 		c.clearServiceState(key, svc)
 	}
 
+	clusterIPs := svc.Spec.ClusterIPs
 	// Clear the lbIP if it has a different ipFamily compared to the clusterIP.
 	// (this should not happen since the "ipFamily" of a service is immutable)
-	if (clusterIP.To4() == nil) != (lbIP.To4() == nil) {
+	if len(clusterIPs) < 2 && len(lbIPs) != 0 {
+		if (clusterIP.To4() == nil) != (lbIPs[0].To4() == nil) {
+			c.clearServiceState(key, svc)
+			lbIPs = nil
+		}
+	} else {
 		c.clearServiceState(key, svc)
-		lbIP = nil
 	}
-
 	// It's possible the config mutated and the IP we have no longer
 	// makes sense. If so, clear it out and give the rest of the logic
 	// a chance to allocate again.
-	if lbIP != nil {
+	if len(lbIPs) != 0 {
 		// This assign is idempotent if the config is consistent,
 		// otherwise it'll fail and tell us why.
-		if err := c.ips.Assign(key, lbIP, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
-			level.Info(l).Log("event", "clearAssignment", "reason", "notAllowedByConfig", "msg", "current IP not allowed by config, clearing")
+		if err = c.ips.Assign(key, lbIPs, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
+			level.Info(l).Log("event", "clearAssignment", "error", err, "msg", "current IP not allowed by config, clearing")
 			c.clearServiceState(key, svc)
-			lbIP = nil
+			lbIPs = nil
 		}
 
 		// The user might also have changed the pool annotation, and
 		// requested a different pool than the one that is currently
 		// allocated.
-		desiredPool := svc.Annotations["metallb.universe.tf/address-pool"]
-		if lbIP != nil && desiredPool != "" && c.ips.Pool(key) != desiredPool {
+		desiredPool := svc.Annotations[annotationAddressPool]
+		if lbIPs != nil && desiredPool != "" && c.ips.Pool(key) != desiredPool {
 			level.Info(l).Log("event", "clearAssignment", "reason", "differentPoolRequested", "msg", "user requested a different pool than the one currently assigned")
 			c.clearServiceState(key, svc)
-			lbIP = nil
+			lbIPs = nil
 		}
-	}
-
-	// User set or changed the desired LB IP, nuke the
-	// state. allocateIP will pay attention to LoadBalancerIP and try
-	// to meet the user's demands.
-	if svc.Spec.LoadBalancerIP != "" && svc.Spec.LoadBalancerIP != lbIP.String() {
-		level.Info(l).Log("event", "clearAssignment", "reason", "differentIPRequested", "msg", "user requested a different IP than the one currently assigned")
+		// User set or changed the desired LB IP, nuke the
+		// state. allocateIP will pay attention to LoadBalancerIP and try
+		// to meet the user's demands.
+		// For dual-stack it will match against the 1st allocated IP.
+		if svc.Spec.LoadBalancerIP != "" && svc.Spec.LoadBalancerIP != lbIPs[0].String() {
```

Comment:
this check is already as part of IPFamilyForAddresses()

Reply:
sorry not following why is that ?

Reply:
because here you only check ip lbIPs[0] matches, but that's not enough I think.
If you have a d/s service you still need to clear because we can't honor it.